### DISCUSSION TITLE
Is IoT/remote capbilities "Device client connection to remote server using popular libraries" still supported? I tried Arduino ESP32 and Ethernet 2 libraries but the code generation seems incomplete?

### DISCUSSION THREAD

**Discussion by AtmosphEng** [davetcc/tcMenu#534] [API & Remote]
When I run tcMenuGenerator I get errors such as:

In file included from src/src_menu.h:6:0,
                 from src\generated\TcmDCC_menu.cpp:14:
src/generated/TcmDCC_menu.h: In function 'MenuItem& rootMenuItem()':
src/generated/TcmDCC_menu.h:38:35: warning: return-statement with no value, in function returning 'MenuItem&' [-fpermissive]
 inline MenuItem& rootMenuItem() { return ; }
                                   ^~~~~~
src\generated\TcmDCC_menu.cpp: In function 'void setupMenu()':
src\generated\TcmDCC_menu.cpp:37:42: error: expected primary-expression before ')' token
     menuMgr.initWithoutInput(&renderer, &);

**Comment by davetcc** [comment]
It should work completely, obviously we can't test every possible combination of library and board, but I've recently tested quite a few of them and they were all working (AVR/Ethernet2, STM32/Stm32Ethernet, ESP32/WiFi, Arduino32/WiFi, mbed/LwIP). 

The best thing to do is to provide the version of TcMenu you're using, an example sketch that recreates the issue, and the either the platformIO configuration, or the Arduino board IDE setup that you used.

**Reply by davetcc** [reply]
Actually looking at the code `menuMgr.initWithoutInput(&renderer, &);` looks suspicious. Like a menuitem with no name

**Reply by AtmosphEng** [reply]
Thanks for the reply and the new 4.4.0 release - I've seen no issues so far, working with PIO. This is the first time I've tried the Device Client Connection (DCC) to remote server code generator option. Initially I hoped a 2nd ESP32 target (e.g. Lilygo t-display-s3) could be used to listen to the 1st ESP32 target and render its tcMenu remotely on the 2nd ESP32 local lcd. If I now understand correctly, do only Java API capable targets like the RPI have this kind of capability? Is there a complete server and client example of how to use DCC local rendering of tcMenu on Github?

**Reply by davetcc** [reply]
Only Java clients can handle the client remote at the moment, it's intended for when you need to aggregate data from many devices on a server somewhere. Just to combine a few menus you could Java APIs menu in menu. 

We're just preparing a new repository now that will be available by the weekend. It includes:

* A complete example of how to build an embedded application on Raspberry PI.
* The original EmbedControl UI with instructions how to build it, should make your life much easier.
* Examples of how to use the Java API with most of the possible permutations covered.
* An example of simple Java web sockets without needing Jetty, using a lighter web socket library. For cases when Apache or other webserver is hosting the static content.

Longer term:

* A couple of examples of the DotNet API showing how to use it.
* A couple of examples of the Python API showing how to use it.
* A UI that will be the approximate equivalent of EmbedControl built using Flutter (iOS and Android)
* Examples of the dart API

**Reply by AtmosphEng** [reply]
Thanks. Sounds good.

**Reply by davetcc** [reply]
https://github.com/TcMenu/tcmenu-examples-starters

**Reply by davetcc** [reply]
with the above repo, you should be able to build the Embed Control UI, customize it to your needs and package it.

**Reply by AtmosphEng** [reply]
Many thanks for the new repo. I'm going to have fun going through it :<)
